README
NAME
    WebService::Geocodio - A Perl interface to Geocod.io

VERSION
    version 0.03

SYNOPSIS
        use 5.014;
        use WebService::Geocodio;
        use WebService::Geocodio::Location;

        my $geo = WebService::Geocodio->new(
            api_key => $ENV{GEOCODIO_API_KEY}
        );

        # Wrigley Field
        my $loc = WebService::Geocodio::Location->new(
            number => 1060,
            postdirection => 'W',
            street => 'Addison',
            suffix => 'Street',
            city => 'Chicago',
            state => 'IL',
        );

        # Could add more than one thing here, even bare strings are OK
        # 20050 = zip code in Washington DC
        $geo->add_location($loc, '20050');

        # prints:
        # Chicago: 41.947205791667, -87.656316875
        # Chicago: 41.947180613636, -87.657167363636
        # Washington: 38.893311, -77.014647

        map { say $_->city, ": ", $_->lat, ", ", $_->lng } $geo->geocode();

OVERVIEW
    This module is a fairly thin wrapper around the Geocod.io
    <http://geocod.io> geocoding web service. This service currently only
    supports US based addresses at the moment. Both forward and reverse
    geocoding is supported.

    In my testing, the service is somewhat finicky about how addresses are
    presented and stored; please read the service API documentation
    thoroughly to make sure you're getting the best quality results from the
    service.

    You will need to obtain a free API key to use this library.

    All errors are fatal and reported by "confess". If you want more
    graceful error handling, you might want to try using Try::Tiny.

ATTRIBUTES
  api_key
    This is the geocod.io API key. It is required.

  locations
    The list of locations you want to geocode. These can be bare strings (if
    you like) or you can use a fancy object like
    WebService::Geocodio::Location which will serialize itself to JSON
    automatically.

  fields
    You may request the following fields be included in the results:

    *   cd

        Congressional District (for the current Congress)

    *   cd113

        Congressional District (for the 113th Congress which runs through
        2015)

    *   stateleg

        The state legislative divisions for this location. The results
        include both House and Senate, unless the location is unicameral
        like Nebraska or Washington D.C., then only a senate result is
        given.

    *   timezone

        The timezone of this location, UTC offset and whether it observes
        daylight saving time.

    *   school

        The unified or elementary/secondary school district identifiers for
        this location.

METHODS
  add_location
    This method takes one or more locations and stores it in the locations
    attribute.

  show_locations
    Show the locations currently set for geocoding.

  clear_locations
    If you want to clear the current list of locations, use this method.

  add_field
    This method takes one or more fields to include in a result set. Valid
    fields are:

    *   cd

    *   cd113

    *   stateleg

    *   timezone

    *   school

    Fields that do not match these valid names are silently discarded.

  geocode
    Send the current list of locations to the geocod.io service.

    Returns undef if there are no locations stored.

    In a list context, returns a list of WebService::Geocodio::Location
    objects. In a scalar context, returns an arrayref of
    WebService::Geocodio::Location objects. The list of objects is presented
    in descending order of accuracy.

  reverse_geocode
    Send the current list of latitude, longitude pairs to the geocod.io
    service.

    Returns undef if there are no locations stored.

    In a list context, returns a list of WebService::Geocodio::Location
    objects. In scalar context, returns an arrayref of
    WebService::Geocodio::Location objects. The list of objects is presented
    in descending order of accuracy.

AUTHOR
    Mark Allen <mrallen1@yahoo.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Mark Allen.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Change history for WebService-Geocodio

0.03        2014-03-25
            Major code refactor 
            Add support for reverse geocoding
            Add rudimentary support for geocod.io fields

0.02        2014-01-23
            Fix testing failure when an api key isn't in the environment.
            There are *no* API changes.
            
0.01        2014-01-22
            Initial release


Makefile.PL

use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A Perl interface to Geocod.io",
  "AUTHOR" => "Mark Allen <mrallen1\@yahoo.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "WebService-Geocodio",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "WebService::Geocodio",
  "PREREQ_PM" => {
    "JSON" => "2.90",
    "Moo::Lax" => "0.15"
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "0.03",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/WebService/Geocodio/JSON.pm
use strict;
use warnings;

package WebService::Geocodio::JSON;
{
  $WebService::Geocodio::JSON::VERSION = '0.03';
}

use Moo::Role;
use JSON;
use Carp qw(confess);


has 'json' => (
    is => 'ro',
    lazy => 1,
    default => sub { JSON->new() },
);


sub encode {
    my ($self, $aref) = @_;

    return $self->json->encode($aref);
}


sub decode {
    my ($self, $data) = @_;

    return $self->json->decode($data);
}

1;

__END__

=pod

=head1 NAME

WebService::Geocodio::JSON

=head1 VERSION

version 0.03

=head1 ATTRIBUTES

=head2 json

A JSON serializer/deserializer instance. Default is L<JSON>.

=head1 METHODS

=head2 encode

Serialize a Perl data structure to a JSON string

=head2 decode

Deserialize a JSON string to a Perl data structure

=head1 AUTHOR

Mark Allen <mrallen1@yahoo.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/WebService/Geocodio/Request.pm
use strict;
use warnings;

package WebService::Geocodio::Request;
{
  $WebService::Geocodio::Request::VERSION = '0.03';
}

use Moo::Role;
use HTTP::Tiny;
use Carp qw(confess);
use WebService::Geocodio::Location;

with 'WebService::Geocodio::JSON';

# ABSTRACT: A request role for Geocod.io


has 'ua' => (
    is => 'ro',
    lazy => 1,
    default => sub { HTTP::Tiny->new(
        agent => "WebService-Geocodio ",
        default_headers => { 'Content-Type' => 'application/json' },
    ) },
);


has 'base_url' => (
    is => 'ro',
    lazy => 1,
    default => sub { 'http://api.geocod.io/v1/' },
);


sub send_forward {
    my $self = shift;

    $self->_request('geocode', $self->encode(@_));
}


sub send_reverse {
    my $self = shift;

    $self->_request('reverse', $self->encode(@_));
}

sub _request {
    my ($self, $op, $content) = @_;

    my $url;
    if ( $self->has_fields ) {
        $url = $self->base_url 
            . "$op?fields=" . join(',', @{ $self->fields }) 
            .  "&api_key=" . $self->api_key
            ;
    }
    else {
        $url = $self->base_url . "$op?api_key=" . $self->api_key;
    }

    my $response = $self->ua->request('POST', $url, { content => $content });

    if ( $response->{success} ) {
        my $hr = $self->decode($response->{content});
        return map { WebService::Geocodio::Location->new($_) } 
            map {; @{$_->{response}->{results}} } @{$hr->{results}};
    }
    else {
        confess "Request to " . $self->base_url . "$op failed: (" . 
            $response->{status} . ") - " . $response->{content};
    }
}


1;

__END__

=pod

=head1 NAME

WebService::Geocodio::Request - A request role for Geocod.io

=head1 VERSION

version 0.03

=head1 ATTRIBUTES

=head2 ua

A user agent object. Default is L<HTTP::Tiny>

=head2 base_url

The base url to use when connecting to the service. Default is 'http://api.geocod.io'

=head1 METHODS

=head2 send_forward

This method POSTs an arrayref of data to the service for processing.  If the
web call is successful, returns an array of L<WebService::Geocodio::Location>
objects.

Any API errors are fatal and reported by C<Carp::confess>.

=head2 send_reverse

This method POSTs an arrayref of data to the service for processing.  If the
web call is successful, returns an array of L<WebService::Geocodio::Location>
objects.

Any API errors are fatal and reported by C<Carp::confess>.

=head1 AUTHOR

Mark Allen <mrallen1@yahoo.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/WebService/Geocodio/Location.pm
use strict;
use warnings;

package WebService::Geocodio::Location;
{
  $WebService::Geocodio::Location::VERSION = '0.03';
}

use Moo::Lax;
use Carp qw(confess);

# ABSTRACT: Location object for use with Geocod.io service.


has [qw(number street suffix postdirection city state zip formatted lat lng accuracy fields)] => (
    is => 'ro',
    predicate => 1,
);


sub BUILDARGS {
    my ( $class, @args ) = @_;

    my $out;
    if (ref($args[0]) eq "HASH") {
        my $hr = $args[0];
        $out->{accuracy} = $hr->{accuracy} if exists $hr->{accuracy};
        $out->{formatted} = $hr->{formatted_address} if exists $hr->{formatted_address};
        $out->{fields} = $hr->{fields} if exists $hr->{fields};
        map { $out->{$_} = $hr->{address_components}->{$_} if exists $hr->{address_components}->{$_} } qw(number street suffix postdirection city state zip);
        map { $out->{$_} = $hr->{location}->{$_} if exists $hr->{location}->{$_} } qw(lat lng);
    }
    elsif ( @args % 2 == 1 ) {
        $out->{formatted} = $args[0];
    }
    else {
        $out = { @args };
    }

    return $out;
}

sub _forward_formatting {
    my $self = shift;

    return $self->formatted if $self->has_formatted;

    if ( ( not $self->has_zip ) && ( not ( $self->has_city && $self->has_state ) ) ) {
        confess "A zip or city-state pair is required.\n";
    }

    my $s;
    if ( $self->has_number && $self->has_street && $self->suffix ) {
        my @f;
        if ( $self->has_postdirection ) {
            @f = qw(number postdirection street suffix);
        }
        else {
            @f = qw(number street suffix);
        }

        $s .= join " ", (map {; $self->$_ } @f);
        $s .= ", ";
    }

    if ( $self->has_zip ) {
        $s .= $self->zip
    }
    else {
        $s .= join ", ", (map {; $self->$_ } qw(city state));
    }

    return $s;
}

sub _reverse_formatting {
    my $self = shift;

    if ( not ( $self->has_lat && $self->has_lng ) ) {
        confess "lat-lng pair is required\n";
    }

    return join ",", ( map {; $self->$_ } qw(lat lng) );
}


1;

__END__

=pod

=head1 NAME

WebService::Geocodio::Location - Location object for use with Geocod.io service.

=head1 VERSION

version 0.03

=head1 ATTRIBUTES

=head2 number

Buiding number

=head2 street

A street name identifier

=head2 postdirection

A address direction like 'SW' or 'S'

=head2 suffix

The type of street 'Ave', 'St', etc.

=head2 city

The city where the streets have no name.

=head2 state

State wherein city is located

=head2 zip

The postal zip code

You B<must> have either a zip code OR a city/state pair.

=head2 lat

The latitude of the location

=head2 lng

The longitude of the location

=head2 accuracy

A float from 0 -> 1 representing the confidence of the lookup results

=head2 formatted

The full address as formatted by the service.

=head2 fields

A hashref to any fields data requested. Will most likely be replaced by field specific
classes in the near future. 

For now, the field name and its subkeys are all naked hash keys. (Sorry. This will
improve.)

=head1 METHODS

=head2 new

The constructor accepts either a bare string OR a list of key/value pairs where
the keys are the attribute names.

=head1 AUTHOR

Mark Allen <mrallen1@yahoo.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/WebService/Geocodio.pm
use strict;
use warnings;

package WebService::Geocodio;
{
  $WebService::Geocodio::VERSION = '0.03';
}

use Moo::Lax;
use Carp qw(confess);
use Scalar::Util qw(blessed);
with('WebService::Geocodio::Request');

# ABSTRACT: A Perl interface to Geocod.io



has 'api_key' => (
    is => 'ro',
    isa => sub { confess "$_[0] doesn't look like a valid api key\n" unless $_[0] =~ /[0-9a-f]+/ },
    required => 1,
);


has 'locations' => (
    is => 'rw',
    default => sub { [] },
);


has 'fields' => (
    is => 'rw',
    predicate => 1,
    default => sub { [] },
);


sub add_location {
    my $self = shift;

    push @{ $self->locations }, @_;
}


sub show_locations {
    my $self = shift;

    return @{ $self->locations };
}


sub clear_locations {
    my $self = shift;

    $self->locations([]);
}


sub add_field {
    my $self = shift;

    push @{ $self->fields }, grep { /cd|cd113|stateleg|timezone|school/ } @_;
}


sub geocode {
    my $self = shift;

    $self->add_location(@_) if scalar @_;

    return undef if scalar @{$self->locations} < 1;

    my @r = $self->send_forward( $self->_format('forward') );

    wantarray ? return @r : return \@r;
}


sub reverse_geocode {
    my $self = shift;

    $self->add_location(@_) if scalar @_;

    return undef if scalar @{$self->locations} < 1;

    my @r = $self->send_reverse( $self->_format('reverse') );

    wantarray ? return @r : return \@r;
}

sub _format {
    my $self = shift;
    my $direction = shift;

    my $method = $direction eq 'forward' ? '_forward_formatting' 
        : '_reverse_formatting';

    return [ map {; blessed $_ ? $_->$method : $_ } @{$self->locations} ];
}

1;

__END__

=pod

=head1 NAME

WebService::Geocodio - A Perl interface to Geocod.io

=head1 VERSION

version 0.03

=head1 SYNOPSIS

    use 5.014;
    use WebService::Geocodio;
    use WebService::Geocodio::Location;

    my $geo = WebService::Geocodio->new(
        api_key => $ENV{GEOCODIO_API_KEY}
    );

    # Wrigley Field
    my $loc = WebService::Geocodio::Location->new(
        number => 1060,
        postdirection => 'W',
        street => 'Addison',
        suffix => 'Street',
        city => 'Chicago',
        state => 'IL',
    );

    # Could add more than one thing here, even bare strings are OK
    # 20050 = zip code in Washington DC
    $geo->add_location($loc, '20050');

    # prints:
    # Chicago: 41.947205791667, -87.656316875
    # Chicago: 41.947180613636, -87.657167363636
    # Washington: 38.893311, -77.014647

    map { say $_->city, ": ", $_->lat, ", ", $_->lng } $geo->geocode();

=head1 OVERVIEW

This module is a fairly thin wrapper around the L<Geocod.io|http://geocod.io>
geocoding web service.  This service currently only supports US based addresses
at the moment.  Both forward and reverse geocoding is supported. 

In my testing, the service is somewhat finicky about how addresses are
presented and stored; please read the service API documentation thoroughly 
to make sure you're getting the best quality results from the service.

You will need to obtain a free API key to use this library.

All errors are fatal and reported by C<confess>.  If you want more graceful
error handling, you might want to try using L<Try::Tiny>.

=head1 ATTRIBUTES

=head2 api_key

This is the geocod.io API key. It is required.

=head2 locations

The list of locations you want to geocode.  These can be bare strings (if you like) or
you can use a fancy object like L<WebService::Geocodio::Location> which will serialize
itself to JSON automatically.

=head2 fields

You may request the following fields be included in the results:

=over 4

=item * cd

Congressional District (for the current Congress)

=item * cd113

Congressional District (for the 113th Congress which runs through 2015)

=item * stateleg

The state legislative divisions for this location. The results include both
House and Senate, unless the location is unicameral like Nebraska or Washington
D.C., then only a senate result is given.

=item * timezone

The timezone of this location, UTC offset and whether it observes daylight
saving time.

=item * school

The unified or elementary/secondary school district identifiers for this location.

=back

=head1 METHODS

=head2 add_location

This method takes one or more locations and stores it in the locations attribute.

=head2 show_locations

Show the locations currently set for geocoding.

=head2 clear_locations

If you want to clear the current list of locations, use this method.

=head2 add_field

This method takes one or more fields to include in a result set. Valid fields are:

=over 4

=item * cd

=item * cd113

=item * stateleg

=item * timezone

=item * school

=back

Fields that do not match these valid names are silently discarded. 

=head2 geocode

Send the current list of locations to the geocod.io service.

Returns undef if there are no locations stored.

In a list context, returns a list of L<WebService::Geocodio::Location> objects.
In a scalar context, returns an arrayref of L<WebService::Geocodio::Location>
objects. The list of objects is presented in descending order of accuracy.

=head2 reverse_geocode

Send the current list of latitude, longitude pairs to the geocod.io service.

Returns undef if there are no locations stored.

In a list context, returns a list of L<WebService::Geocodio::Location> objects.
In scalar context, returns an arrayref of L<WebService::Geocodio::Location> 
objects.  The list of objects is presented in descending order of accuracy.

=head1 AUTHOR

Mark Allen <mrallen1@yahoo.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
